### DISCUSSION TITLE
Convert std::set to frozenset

### DISCUSSION THREAD

**Discussion by thony-murray** [pybind/pybind11#3836] [Q&A]
I am trying to convert the type `std::map<std::set<int>, int>` from C++ to Python. All of these types have built-in conversions, but the conversion fails with the error `TypeError: unhashable type: 'set'`. This makes sense, as the built-in conversions are tying to create the invalid type `Dict[Set[int], int]` in python. Is there any way I can get it to convert my `std::set` to a Python `frozenset` instead? `frozenset` is hashable and can be used as map keys.

EDIT: The conversion in the reverse direction is desirable as well. If I have a C++ function that takes a `std::map<std::set<int>, int>` I'd like to be able to bind it to Python and pass in a `Dict[FrozenSet[int], int]`.

**Comment by jiwaszki** [comment]
**Update:** actually I have prototyped something like this. I will test it out more and probably create a PR with this. For now feel free to use this sample in your code. I will be happy to hear some feedback!

Here is my gist with answer:
https://gist.github.com/jiwaszki/5ab6267672929e92e6542a44a27fdaa9

And results from tests:
```python
frozenset({1, 2, 3})
frozenset({1, 2, 3})
frozenset({5, 6, 7})
frozenset({5, 6, 7})
{frozenset({1, 2, 3}): 6}
{frozenset({1, 2, 3}): 6}
{1, 2, 3}
{1, 2, 3}
{5, 6, 7}
{5, 6, 7}
Traceback (most recent call last):
  File "./test_frozenset.py", line 13, in <module>
    print(mymodule.foo5({{1,2,3}: 6}))
TypeError: unhashable type: 'set'
```

So to create `py::frozenset` that is fully functional like `py::set` (that translates to `std::set`) you need to have two major parts of code. First is implementation of Python type that uses Python C API, this is easy part. I used modified version of `py::set` that is working for `frozenset` instead.
Now going to problematic part. ** Best case scenario is to catch both cases `std::set<const type>` (or even `const std::set<const type>`) and `std::set<type>`. Why is it important? It should be possible to distinguish `frozenset` and `set`.**

However now there is an issue that pybind maintainers should answer: is adding new class based on top of already covered C++ type is something that breaks existing compatibility between versions? I guess I need to find some answers... 😃

**Reply by thony-murray** [reply]
Thanks so much for working on this!! I don't have time to dig into your code right now but I will try to take a look this weekend if I have some time

**Reply by Skylion007** [reply]
Adding a new caster or pytype does not break compatibility between versions. Even if it did, we do allow ABI breaks for good reason.

**Reply by ecatmur** [reply]
I hope you don't mind - I needed this so I've expanded it into a PR handling sequence types also (as Tuple[T, ...]): https://github.com/pybind/pybind11/pull/3886

**Reply by jiwaszki** [reply]
@ecatmur Not at all! I was busy last two weeks, that is great to see someone expanding on my idea! 😄

**Comment by Skylion007** [comment]
@jiwaszki If you want an example, Eigen::Ref has slightly different behavior for `Eigen::Ref<MatrixType>` and `Eigen::Ref<const MatrixType`https://pybind11.readthedocs.io/en/stable/advanced/cast/eigen.html#pass-by-reference

**Reply by jiwaszki** [reply]
Thank you! I will dig into that in a meantime.

**Reply by Skylion007** [reply]
@jiwaszki More specifically, here is the check: https://github.com/pybind/pybind11/blob/c4e295287b74c0b3f0f8a304cdb746ac2bcb37ad/include/pybind11/eigen.h#L273
